### ISSUE TITLE
Add accessibility (role/aria-*) to view engine

### ISSUE THREAD

**Issue by danieljsummers** [giraffe-fsharp/Giraffe#290]
This is the required pre-PR clearance of a feature I'd like to add. If you think it's too heavy for Giraffe proper, I'll spin it out as a separate package; just let me know.

The W3C [WIA-ARIA specification](https://www.w3.org/TR/wai-aria-1.1/) lays out [attributes (summarized in the HTML spec)](http://w3c.github.io/html/infrastructure.html#element-attrdef-aria-role) that can be applied to various elements that will allow screen readers and other devices to more richly interpret the intent of what an item is supposed to do. It's an area of web development that doesn't receive a lot of attention (though that's changing some), but I believe it's one that we should encourage. Providing the change I'm proposing will do that.

I propose adding an `Accessibility` module (which could be opened), that will contain:
 - A `_role` attribute
 - A `Role` module (under `Accessibility`, requiring qualified access) that is either a DU or a series of `let` statements enumerating the values at the above "attributes" link; the `_role` attribute will take one of these as its lone argument, and will be rendered into the document as its proper value
 - An `_aria` attribute
 - An `Aria` module (under `Accessibility`, requiring qualified access) that is either a DU or a series of `let` statements enumerating the values at the above "attributes" link; the `_aria` attribute will take one of these as its first argument, using it to render the * part of the `aria-*` attribute, and a string value (just like other attributes)

Putting these under an separate module would a) prevent them from appearing in/polluting the workspace unless explicitly opened; and b) allow them to be moved out/around as required. Many role/aria values conflict with HTML names, so requiring qualified access to those modules will eliminate this as a source of name collisions.

(Also, consider whether this would be a good structure for the `_on*` attributes, which I'm guessing are mostly missing because of a design decision. They could go under an `Events` module without polluting the current space (unless the module was opened), though they wouldn't need the enumerated value modules.)

I've got this coded already; here's what it looks like...

```fsharp
/// Attributes to support WAI-ARIA accessibility guidelines
// (values for area- and role attributes obtained from
//  http://w3c.github.io/html/infrastructure.html#element-attrdef-aria-role)
module Accessibility =
    
    /// See Aria module for valid attributes
    let _aria ``type`` = attr (sprintf "aria-%s" ``type``)
    /// See Role module for valid values
    let _role          = attr "role"

    /// Valid role attribute content values
    [<RequireQualifiedAccess>]
    module Role =
        let alert         = "alert"
        let alertdialog   = "alertdialog"
        let application   = "application"
        let article       = "article"
        let banner        = "banner"
        let button        = "button"
        let cell          = "cell"
        let checkbox      = "checkbox"
        let columnheader  = "columnheader"
        let combobox      = "combobox"
        let complementary = "complementary"
        let contentinfo   = "contentinfo"
        let definition    = "definition"
        let dialog        = "dialog"
        let directory     = "directory"
        let document      = "document"
        let feed          = "feed"
        let figure        = "figure"
        let form          = "form"
        let grid          = "grid"
        let gridcell      = "gridcell"
        let group         = "group"
        let heading       = "heading"
        let img           = "img"
        let link          = "link"
        let list          = "list"
        let listbox       = "listbox"
        let listitem      = "listitem"
        let log           = "log"
        let main          = "main"
        let marquee       = "marquee"
        let math          = "math"
        let menubar       = "menubar"
        let navigation    = "navigation"
        let none          = "none"
        let note          = "note"
        let option        = "option"
        let presentation  = "presentation"
        let progressbar   = "progressbar"
        let radio         = "radio"
        let radiogroup    = "radiogroup"
        let region        = "region"
        let row           = "row"
        let rowgroup      = "rowgroup"
        let rowheader     = "rowheader"
        let scrollbar     = "scrollbar"
        let search        = "search"
        let searchbox     = "searchbox"
        let separator     = "separator"
        let slider        = "slider"
        let spinbutton    = "spinbutton"
        let status        = "status"
        let switch        = "switch"
        let tab           = "tab"
        let table         = "table"
        let tablist       = "tablist"
        let tabpanel      = "tabpanel"
        let term          = "term"
        let textbox       = "textbox"
        let timer         = "timer"
        let toolbar       = "toolbar"
        let tooltip       = "tooltip"
        let tree          = "tree"
        let treegrid      = "treegrid"
        let treeitem      = "treeitem"
    
    /// Valid aria attributes
    [<RequireQualifiedAccess>]
    module Aria =
        let activedescendant = "activedescendant"
        let atomic           = "atomic"
        let autocomplete     = "autocomplete"
        let busy             = "busy"
        let ``checked``      = "checked"
        let colcount         = "colcount"
        let colindex         = "colindex"
        let colspan          = "colspan"
        let controls         = "controls"
        let current          = "current"
        let describedby      = "describedby"
        let details          = "details"
        let dialog           = "dialog"
        let disabled         = "disabled"
        let dropeffect       = "dropeffect"
        let errormessage     = "errormessage"
        let expanded         = "expanded"
        let flowto           = "flowto"
        let grabbed          = "grabbed"
        let haspopup         = "haspopup"
        let hidden           = "hidden"
        let invalid          = "invalid"
        let keyshortcuts     = "keyshortcuts"
        let label            = "label"
        let labelledby       = "labeledby"
        let level            = "level"
        let live             = "live"
        let multiline        = "multiline"
        let multiselectable  = "multiselectable"
        let orientation      = "orientation"
        let owns             = "owns"
        let placeholder      = "placeholder"
        let posinset         = "posinset"
        let pressed          = "pressed"
        let readonly         = "readonly"
        let relevant         = "relevant"
        let required         = "required"
        let roledescription  = "roledescription"
        let rowcount         = "rowcount"
        let rowindex         = "rowindex"
        let rowspan          = "rowspan"
        let selected         = "selected"
        let setsize          = "setsize"
        let sort             = "sort"
        let valuemax         = "valuemax"
        let valuemin         = "valuemin"
        let valuenow         = "valuenow"
        let valuetext        = "valuetext"
```

Thanks!

**Comment by dustinmoris** [comment]
Hi,

First of all thanks for the great explanation and the code examples and the ambition to add this functionality into Giraffe! I really like it!

I like your suggestion of having these new attributes under a new module, but I would probably nest it under the current `Giraffe.GiraffeViewEngine` module, so that the full namespace ends up being `Giraffe.GiraffeViewEngine.Accessibility` (or similar).

In terms of implementation, I had a thought and personally I think I'd prefer to have unambiguous functions for the roles and aria elements.

Example:

Instead of:

```
let _role = attr "role"

[<RequireQualifiedAccess>]
    module Role =
        let alert = "alert"
```

I would suggest:

```
let _roleAlert = attr "role" "alert"
```

This would have the benefit that

a.) the default Giraffe programming model wouldn't allow a user to generate an invalid role
b.) there would be no disconnect between the `_role` attribute and the list of allowable values
c.) all valid role attributes could be easily discovered in an IDE by simply starting to type `_role...` which would bring up the intellisense and suggest a list of autocompletion values

The same I would apply for the `aria-{...}` attributes:

```
let _ariaActiveDescendant = attr "aria-activedescendant"
```

Otherwise I think this sounds great and would definitely love you to submit a PR!

Let me know what you think of my suggestions and feel free to counter argue if you think there's a benefit of doing it in a different way ;)!

**Comment by danieljsummers** [comment]
Excellent. I'm fine with the suggested change; I should have this submitted by this weekend.

**Comment by dustinmoris** [comment]
Awesome, thanks!

**Comment by dustinmoris** [comment]
This has been released with [Giraffe 2.0.0](https://github.com/giraffe-fsharp/Giraffe/releases/tag/v2.0.0).
